src/workerd/api/sockets.c++

```diff
@@ -124,6 +164,30 @@ jsg::Promise<void> Socket::close(jsg::Lock& js) {
   }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
 }
 
+jsg::Ref<Socket> Socket::startTls(jsg::Lock& js, jsg::Optional<TlsOptions> tlsOptions) {
+  JSG_REQUIRE(!isSecureSocket, TypeError, "Cannot startTls on a TLS socket.");
+
+  // The current socket's writable buffers need to be flushed. The socket's WritableStream is backed
+  // by an AsyncIoStream which doesn't implement any buffering, so we don't need to worry about
+  // flushing. But this is something to keep in mind in case this assumption no longer holds in
+  // the future.
+  //
+  // Detach the AsyncIoStream from the Writable/Readable streams and make them unusable.
+  writable->removeSink(js);
+  readable->detach(js);
+  closeFulfiller.resolver.resolve();
+
+  auto acceptedHostname = kj::heapString(domain);
+  KJ_IF_MAYBE(s, tlsOptions) {
+    KJ_IF_MAYBE(expectedHost, s->expectedServerHostname) {
+      acceptedHostname = kj::heapString(*expectedHost);
```

Comment:
I don't think either string allocation is necessary -- we should be able to just pass the StringPtrs to `tlsStarter()`, which uses them synchronously.